Quantized H Tangent function

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31031 Quantized H Tangent function**

This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D19334280](https://our.internmc.facebook.com/intern/diff/D19334280)

Commit:
[Draft] Quantized H Tangent function

This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

z-a-f commented:
Didn't run the tests yet -- still compiling, hence "draft" :)

ljk53 commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -192,6 +192,42 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+      iter,
+      [&](scalar_t value_qx) -> scalar_t {
+        const auto value_dx = at::dequantize_val(scale, zero_point, value_qx);
+        return at::quantize_val(scale, zero_point, tanh(value_dx));
```
what tanh implementation does it use? shall we explicitly call std::tanh?

ljk53 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -2636,6 +2636,10 @@
   variants: function, method
+  dispatch:
+    CPU: tanh
+    CUDA: tanh
```
do we need CUDA here? check other nearby entries CPU/CUDA all point to different implementation...

z-a-f replied:
I am not 100% familiar with the `native_functions.yaml`. Originally it didn't have explicit `dispatch`, which (I assumed) would point to default `tanh` implementation for all devices. However, because I am adding `quantized_tanh`, I have to specify the `dispatch` explicitly. That's why I put the both the CPU and CUDA in there

raghuramank100 replied:
please test to ensure that cuda functionality for tanh is also tested in this case after the changes

Commit:
Update on "[Draft] Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

z-a-f commented on aten/src/ATen/native/quantized/cpu/qtanh.cpp:
```diff
@@ -0,0 +1,88 @@
+    input.options(),
+    input_contig.q_scale(),
+    input_contig.q_zero_point());
+  size_t num_elems_y = input_contig.numel() / qy.size(0);
```
That's a good point -- they should be the same

z-a-f replied:
keeping it for discussion and consistency for now -- will change before landing

Commit:
Update on "[Draft] Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

z-a-f commented:
All tests pass locally

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

kostmo commented:
## CircleCI build failures summary
As of commit 44467e9:
* **1/1** broken upstream at merge base 7a8261e (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=7a8261e962f26f84f6211bca4270713a3102146c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=44467e91495ea8c0a0ce6d16ff7ff66b503b616b).

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_xla_linux_xenial_py3_6_clang7_test](https://dr.pytorch.org/build-details.html?build_id=72802529)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

supriyar commented on test/test_quantized.py:
```diff
@@ -144,6 +144,28 @@ def test_qrelu6(self, X):
                 self.assertEqual(qY, qY_hat,
                                  message="{} relu failed".format(name))
 
+    """Tests the correctness of the quantized::relu op."""
```
relu -> tanh

supriyar commented on test/test_quantized.py:
```diff
@@ -144,6 +144,28 @@ def test_qrelu6(self, X):
+    """Tests the correctness of the quantized::relu op."""
+    @given(X=hu.tensor(shapes=hu.array_shapes(1, 5, 1, 5),
+                       qparams=hu.qparams()))
+    def test_qtanh(self, X):
```
Can you add testing of the qnnpack op as well? You can add the `qengine` as a hypothesis param, similar to what we did for qlinear/qconv.

supriyar reviewed (changes requested):
Mainly missing testing for qnnpack tanh.

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:
```diff
@@ -192,6 +193,42 @@ void qrelu6_kernel(const Tensor& qx, Tensor& qy) {
+      [&](Vec value_qx) -> Vec {
+        const auto value_dx = value_qx.dequantize(scale_vec, zero_point_vec, scale_neg_zp_premul_vec);
+        Vec::float_vec_return_type retvals;
+        for (int idx = 0; idx < Vec::float_num_vecs(); ++idx) {
```
Lets go for a quantized tanh implementation and remove the need to call float tanh here for speed. (@hx89: Do we have a look up table based implementation ready?).
We should eventually have a lookup table that takes in an input in the range [-4,4] and returns values in the range  [-1 ,1]. tanh(4) = 0.9993. The table itself is 8 bits wide. We will need to determine the size of the table via experiments. This can occur in a later PR if we dont have a quantized implementation ready.

z-a-f replied:
I don't think the float here has a huge implication, as in the end of the day it uses the native `tanh` implementation. But, you are probably right -- the LUT will be faster. However, I don't think it will be scalable -- how are we going to handle the `qint32` or `qint16` down the road?

hx89 replied:
Yes @raghuramank100 , we have LUT based implementation for tanh in DNNLOWP: https://github.com/pytorch/pytorch/blob/master/caffe2/quantization/server/tanh_dnnlowp_op.cc

hx89 replied:
Yes, we have LUT implementation of tanh in DNNLOWP:
https://github.com/pytorch/pytorch/blob/master/caffe2/quantization/server/tanh_dnnlowp_op.cc
If higher accuracy is needed we can use Taylor series to add more terms. We have an internal LUT implementation for float16 tanh using Taylor series.

hx89 replied:
Yes, we have LUT implementation of tanh in DNNLOWP:
https://github.com/pytorch/pytorch/blob/master/caffe2/quantization/server/tanh_dnnlowp_op.cc
If higher accuracy is needed we can use Taylor series to add more terms. We have an internal LUT implementation for float16 tanh using Taylor series.

z-a-f replied:
How does LUT (8 and 16 bit) compare to `std::tanh`?

jerryzh168 replied:
I think we can optimize later, I don't feel this would be a bottleneck of the network

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/tanh.c:
```diff
@@ -108,7 +108,7 @@ enum pytorch_qnnp_status pytorch_qnnp_create_tanh_nc_q8(
         input_scale * (float)(i - (int32_t)(uint32_t)input_zero_point);
     /* Scale tanh(x) by 1 / output scale = 256.0 */
-    float scaled_tanh_x = 256.0f * tanh(x);
+    float scaled_tanh_x = 128.0f * tanh(x) + 128.0f;
```
Do the qnnpack tests for tanh pass with this change (aten/src/ATen/native/quantized/cpu/qnnpack/test/tanh-operator-tester.h)? I'm guessing that needs to change as well.

z-a-f replied:
This is the new scale -- I just forgot to change the comment. The scale is `2.0 / 256 = 1.0/128`, and the zero_point is 128 -- I am assuming the UINT8. I will update the comment

supriyar replied:
I was referring to the unit tests here -https://github.com/pytorch/pytorch/blob/cc319659e387634359c2f188ea3adefeb2bc2954/aten/src/ATen/native/quantized/cpu/qnnpack/test/tanh-operator-tester.h#L93. I believe the scale and zero point for output ref needs to be updated here as well.

z-a-f replied:
Oh my bad -- made the change, lemme test

z-a-f replied:
```console
$> ./scripts/build_local.sh -DBUILD_BINARY=ON -DBUILD_PYTHON=OFF -DUSE_OBSERVERS=OFF -DUSE_DISTRIBUTED=OFF
$> ./build/local/tanh-test
Running main() from gtest_main.cc
[==========] Running 18 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 18 tests from TANH_OP
[ RUN      ] TANH_OP.zero_batch
[       OK ] TANH_OP.zero_batch (0 ms)
[ RUN      ] TANH_OP.unit_batch
[       OK ] TANH_OP.unit_batch (1 ms)
[ RUN      ] TANH_OP.unit_batch_with_qmin
[       OK ] TANH_OP.unit_batch_with_qmin (0 ms)
[ RUN      ] TANH_OP.unit_batch_with_qmax
[       OK ] TANH_OP.unit_batch_with_qmax (0 ms)
[ RUN      ] TANH_OP.unit_batch_with_input_scale
[       OK ] TANH_OP.unit_batch_with_input_scale (1 ms)
[ RUN      ] TANH_OP.unit_batch_with_input_zero_point
[       OK ] TANH_OP.unit_batch_with_input_zero_point (1 ms)
[ RUN      ] TANH_OP.small_batch
[       OK ] TANH_OP.small_batch (0 ms)
[ RUN      ] TANH_OP.small_batch_with_input_stride
[       OK ] TANH_OP.small_batch_with_input_stride (1 ms)
[ RUN      ] TANH_OP.small_batch_with_output_stride
[       OK ] TANH_OP.small_batch_with_output_stride (0 ms)
[ RUN      ] TANH_OP.small_batch_with_qmin
[       OK ] TANH_OP.small_batch_with_qmin (0 ms)
[ RUN      ] TANH_OP.small_batch_with_qmax
[       OK ] TANH_OP.small_batch_with_qmax (0 ms)
[ RUN      ] TANH_OP.small_batch_with_input_scale
[       OK ] TANH_OP.small_batch_with_input_scale (1 ms)
[ RUN      ] TANH_OP.small_batch_with_input_zero_point
[       OK ] TANH_OP.small_batch_with_input_zero_point (1 ms)
[ RUN      ] TANH_OP.strided_batch
[       OK ] TANH_OP.strided_batch (1 ms)
[ RUN      ] TANH_OP.strided_batch_with_qmin
[       OK ] TANH_OP.strided_batch_with_qmin (0 ms)
[ RUN      ] TANH_OP.strided_batch_with_qmax
[       OK ] TANH_OP.strided_batch_with_qmax (0 ms)
[ RUN      ] TANH_OP.strided_batch_with_input_scale
[       OK ] TANH_OP.strided_batch_with_input_scale (2 ms)
[ RUN      ] TANH_OP.strided_batch_with_input_zero_point
[       OK ] TANH_OP.strided_batch_with_input_zero_point (1 ms)
[----------] 18 tests from TANH_OP (10 ms total)

[----------] Global test environment tear-down
[==========] 18 tests from 1 test case ran. (10 ms total)
[  PASSED  ] 18 tests.
```

z-a-f replied:
I am not 100% sure why it was passing before, honestly. I think the reason is that the reference is also using the same scale/zp as the implementation, so it has the same output range. The scale only affects the range, when the input is relatively large. I think what is happening is that given the scale/zp the reference produced the same results as the qtanh

supriyar commented on aten/src/ATen/native/quantized/cpu/qtanh.cpp:
```diff
@@ -0,0 +1,88 @@
+DEFINE_DISPATCH(qtanh_stub);
+
+#ifdef USE_PYTORCH_QNNPACK
+// This ALWAYS outputs scale=2.0/256, zp=128, dtype=quint8
```
Did this change from 1/256.0 and zp=0 because of the symmetric range between [-1, 1]? Curious as to how the tests were passing earlier then

z-a-f replied:
There were no tests for the QNNpack -- I enabled them after your previous comment. After implementing the tests, this was caught

supriyar commented on test/test_quantized.py:
```diff
@@ -114,6 +114,27 @@ def test_qrelu(self, X):
+                       qparams=hu.qparams()))
+    def test_qtanh(self, X):
+        # QNNPACK is tested separately in TestQNNPackOps
+        if torch.backends.quantized.engine == 'qnnpack':
```
This should not be necessary as by default engine is not set to qnnpack.

supriyar commented on test/test_quantized.py:
```diff
@@ -114,6 +114,27 @@ def test_qrelu(self, X):
+        qY = torch.quantize_per_tensor(Y, scale=scale, zero_point=zero_point,
+                                       dtype=torch_type)
+        qY_hat = torch.tanh(qX)
+        self.assertEqual(qY, qY_hat, message="QNNPACK TanH failed!")
```
Message needs to be updated here.

supriyar commented on test/test_quantized.py:
```diff
@@ -1809,6 +1830,27 @@ def test_qnnpack_relu(self, X):
+                       qparams=hu.qparams(dtypes=torch.quint8)))
+    def test_qnnpack_tanh(self, X):
+        # Note: In QNNPACK the output scale and zero_point can be only
+        #       1.0/256, 0 respectively, because it uses an LUT with 256 bins.
```
I think based on the new scale, zp this comment needs to be updated?

supriyar commented on test/test_quantized.py:
```diff
@@ -1809,6 +1830,27 @@ def test_qnnpack_relu(self, X):
+                                           dtype=torch.quint8)
+            qY_hat = torch.tanh(qX)
+
+            self.assertEqual(qY, qY_hat, message="QNNPACK TanH failed!")
```
Should we also compare the results between QNNPACK and server-side tanH implementation?

z-a-f replied:
What do you mean? This test is doing exactly that.

supriyar replied:
Yes, I meant server side quantized tanh implementation. Since one of them uses LUT and other does not, I thought it would good to compare them to make sure they don't diverge

z-a-f replied:
I don't think there is a point -- if both of them are compared to the FP reference, and both of them are within the tolerance, I am not sure if comparing the two would catch anything that FP ref does not.

But if you think there is still need for it -- it's an easy thing to do

z-a-f replied:
Added the tests -- also, rewrote the implementation of the naïve implementation to always have the same output scale/zero_point. That way the accuracies can match across platforms

supriyar reviewed (changes requested):
Please see comments

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

supriyar commented on test/test_quantized.py:
```diff
@@ -1809,6 +1838,32 @@ def test_qnnpack_relu(self, X):
+        Y = torch.tanh(X)
+        qY = torch.quantize_per_tensor(Y, scale=1.0 / 128, zero_point=128,
+                                       dtype=torch.quint8)
+        with override_quantized_engine('fbgemm'):
```
You don't need an override here. It should call the server side implementation of quantized tanh by default. 
Can you also call it something other than fbgemm? (`qYserver` maybe)

supriyar reviewed (approved):
LGTM - one last comment needs to be addressed. Thanks!

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)

izdeby commented:
@z-a-f, fyi, i had to revert this PR as it have introduced a build failure on pytorch_macos_10_13_py3_test job.

Failure: 
`
Jan 08 19:23:52 Running test_quantized ... [2020-01-08 19:23:52.452890]
Jan 08 19:23:52 Traceback (most recent call last):
Jan 08 19:23:52   File "test_quantized.py", line 2, in <module>
Jan 08 19:23:52     from future.builtins import round
Jan 08 19:23:52 ModuleNotFoundError: No module named 'future'
Jan 08 19:23:52 Traceback (most recent call last):
Jan 08 19:23:52   File "test/run_test.py", line 456, in <module>
Jan 08 19:23:52     main()
Jan 08 19:23:52   File "test/run_test.py", line 449, in main
Jan 08 19:23:52     raise RuntimeError(message)
Jan 08 19:23:52 RuntimeError: test_quantized failed!
`

Commit:
Update on "Quantized H Tangent function"


This activation will be needed for the LSTM implementation.
Also includes the QNNPack implementation.

Differential Revision: [D18903453](https://our.internmc.facebook.com/intern/diff/D18903453)
